### PR TITLE
FS Auto mount changes

### PR DESCRIPTION
**Pull request by jeongchanKim**

### CODE REVIEW

File: os/board/common/Kconfig [hunk 1]
```diff
@@ -106,3 +106,27 @@ config SECOND_FLASH_PART_NAME
...
+	bool "Automount userfs partiton"
+	default n
+	depends on AUTOMOUNT
+	depends on FS_SMARTFS
+	---help---
+		If enabled, mount userfs partition automatically at boot.
+
+config AUTOMOUNT_ROMFS
+	bool "Automount romfs partiton"
+	default y
```
Comment (sunghan-chang) [inline os/board/common/Kconfig:128]: Is there any reason to be y even AUTOMOUNT_USERFS is defaultly n?

File: os/board/stm32l4r9ai-disco/src/stm32_exflash.c [hunk 2]
```diff
@@ -126,27 +127,18 @@ void stm32_exflash_initialize(void)
 #if defined(CONFIG_FLASH_PARTITION)
     int ret;
     struct mtd_dev_s *mtd;
+    partition_info_t partinfo;
 
     mtd = (FAR struct mtd_dev_s *)mtd_initialize();
-    ret = configure_partitions(mtd, &g_flash_part_data);
+    ret = configure_partitions(mtd, &g_flash_part_data, &partinfo);
```
Comment (amogh-samsung) [inline os/board/stm32l4r9ai-disco/src/stm32_exflash.c:133]: This needs to be changed to `configure_mtd_partitions` please.
Causes problem in the current STM32L4r9ai-disco builds as well.

File: os/board/common/partitions.c [hunk 3]
```diff
@@ -227,4 +237,38 @@ int configure_mtd_partitions(struct mtd_dev_s *mtd, struct partition_data_s *par
 
 	return OK;
 }
+
+void automount_fs_partition(partition_info_t *partinfo)
```
Comment (amogh-samsung) [inline os/board/common/partitions.c:241]: The main reason for these changes is generalization of userfs and romfs mounting code?
Or is it also intended to serve some other specific purpose?

File: os/board/common/Kconfig (hunk 1 above)
Comment (jeongchanKim) [reply os/board/common/Kconfig:128]: Changed. Thanks.

File: os/board/common/partitions.c (hunk 3 above)
Comment (jeongchanKim) [reply os/board/common/partitions.c:241]: It is for generalization of userfs and romfs.
Do you have any idea?

File: os/board/stm32l4r9ai-disco/src/stm32_exflash.c (hunk 2 above)
Comment (jeongchanKim) [reply os/board/stm32l4r9ai-disco/src/stm32_exflash.c:133]: Done. Thanks.

File: os/board/common/partitions.c [hunk 4]
```diff
@@ -78,40 +80,48 @@ FAR struct mtd_dev_s *mtd_initialize(void)
...
-				lldbg("ERROR: failed to initialise mtd ftl errno :%d\n", errno);
-				return;
-			}
-		} else
+	) {
+		if (ftl_initialize(partno, mtd_part)) {
+			lldbg("ERROR: failed to initialise mtd ftl errno :%d\n", errno);
+			return;
+		} else {		
+			partinfo->romfs_partno = partno;
```
Comment (sunghan-chang) [inline os/board/common/partitions.c:102]: This line is NOT only for romfs. This will cause overwrite issue.

File: os/board/imxrt1020-evk/Kconfig [hunk 5]
```diff
@@ -70,31 +35,6 @@ config IMXRT_AUTOMOUNT_SSSRW_MOUNTPOINT
...
-
-config IMXRT_AUTOMOUNT_ROMFS_MOUNTPOINT
-	string "Mountpoint of the partition for romfs file system"
-	default "/rom"
-	depends on IMXRT_AUTOMOUNT_ROMFS
-	---help---
-		Specifies the mount point where romfs file system
-		will be mounted at.
-
 config IMXRT_AUTOMOUNT_TZDEVNAME
```
Comment (sunghan-chang) [inline os/board/imxrt1020-evk/Kconfig:38]: For next step, please check this to merge into new common function.

Comment (jeongchanKim) [reply os/board/imxrt1020-evk/Kconfig:38]: I will check and make another PR if needed.

File: os/board/common/partitions.c (hunk 4 above)
Comment (jeongchanKim) [reply os/board/common/partitions.c]: Done. Thanks.

**Comment by sunghan-chang** [comment]
Please resolve the conflict.

File: os/board/artik05x/src/artik05x_tash.c [hunk 6]
```diff
@@ -171,36 +171,18 @@ int board_app_initialize(void)
...
 
 	mtd = (FAR struct mtd_dev_s *)mtd_initialize();
 	/* Configure mtd partitions */
-	configure_mtd_partitions(mtd, &g_flash_part_data);
+	configure_mtd_partitions(mtd, &g_flash_part_data, &partinfo);
 
-#ifdef CONFIG_ARTIK05X_AUTOMOUNT
-#ifdef CONFIG_ARTIK05X_AUTOMOUNT_USERFS
-	/* Initialize and mount user partition (if we have) */
-#ifdef CONFIG_SMARTFS_MULTI_ROOT_DIRS
```
Comment (sunghan-chang) [inline os/board/artik05x/src/artik05x_tash.c:186]: This is disappeared in new function. Could you check this?

File: os/board/common/Kconfig [hunk 7]
```diff
@@ -106,3 +106,27 @@ config SECOND_FLASH_PART_NAME
...
+	depends on AUTOMOUNT
+	depends on FS_SMARTFS
+	---help---
+		If enabled, mount userfs partition automatically at boot.
+
+config AUTOMOUNT_ROMFS
+	bool "Automount romfs partiton"
+	default n
+	depends on AUTOMOUNT
+	depends on FS_ROMFS
```
Comment (jeongarmy) [inline os/board/common/Kconfig:130]: TMPFS supports automount also. How about adding AUTOMOUNT_TMPFS here?

Comment (jeongchanKim) [reply os/board/common/Kconfig:130]: It has already common function, so I will not care about it in this PR.
If needed, make new PR for merging it sometime after.

**Comment by Taejun-Kwon** [comment]
This shows how developer can start file system. welcome =)

File: os/board/artik05x/src/artik05x_tash.c (hunk 6 above)
Comment (jeongchanKim) [reply os/board/artik05x/src/artik05x_tash.c:186]: Done. Thanks.

File: os/board/common/partitions.c [hunk 8]
```diff
@@ -78,40 +81,62 @@ FAR struct mtd_dev_s *mtd_initialize(void)
...
-
+#ifdef CONFIG_MTD_FTL
+	if (do_ftlinit) {
+		if (ftl_initialize(partno, mtd_part)) {
+			lldbg("ERROR: failed to initialise mtd ftl errno :%d\n", errno);
+			return;
+		}
+#ifdef CONFIG_FS_ROMFS
+		if (save_romfs_partno) {
+				partinfo->romfs_partno = partno;
```
Comment (sunghan-chang) [inline os/board/common/partitions.c:135]: Could you remove a tab for alignment?

File: os/board/common/partitions.c [hunk 9]
```diff
@@ -78,40 +81,62 @@ FAR struct mtd_dev_s *mtd_initialize(void)
...
+		partinfo->smartfs_partno = partno;
+	}
 #endif
-		{
-
+#ifdef CONFIG_MTD_FTL
+	if (do_ftlinit) {
+		if (ftl_initialize(partno, mtd_part)) {
+			lldbg("ERROR: failed to initialise mtd ftl errno :%d\n", errno);
+			return;
```
Comment (sunghan-chang) [inline os/board/common/partitions.c:131]: How about automount for romfs with this case?

File: os/board/common/partitions.c (hunk 3 above)
Comment (amogh-samsung) [reply os/board/common/partitions.c:262]: We have to scan the partition list in config_mtd_partitions to get "partno" and "minor".
We also need to do that when we use dump_tool to extract primary or external flash partition.

Comment (jeongchanKim) [reply os/board/common/partitions.c:262]: If you want to expand this function, you can do after merging this.
If possible to use common, it is good for most of the cases.

File: build/configs/artik053/minimal/defconfig [hunk 10]
```diff
@@ -202,7 +202,7 @@ CONFIG_BOOT_RUNFROMFLASH=y
 #
 # Boot Memory Configuration
 #
-CONFIG_RAM_KREGIONx_START="0x02023800"
+CONFIG_RAM_KREGIONx_START="0x02023800	"
```
Comment (sunghan-chang) [inline build/configs/artik053/minimal/defconfig:205]: Let's revert this.

File: build/configs/artik053/tc_minimal_kernel/defconfig [hunk 11]
```diff
@@ -247,6 +247,7 @@ CONFIG_FLASH_PARTITION=y
 CONFIG_FLASH_MINOR=0
 CONFIG_FLASH_PART_SIZE="16,48,192,32,512,2400,1536,1536,1400,8,512,"
 CONFIG_FLASH_PART_TYPE="none,ftl,none,none,none,none,none,ftl,smartfs,config,none,"
+# CONFIG_AUTOMOUNT is not set
```
Comment (sunghan-chang) [inline build/configs/artik053/tc_minimal_kernel/defconfig:250]: Need to remove line 240

File: os/board/common/partitions.c [hunk 12]
```diff
@@ -36,11 +37,12 @@
 #endif
 #include "common.h"
 
+#define FS_PATH_MAX 15
+
 #ifdef CONFIG_FLASH_PARTITION
 struct partition_data_s g_flash_part_data = {
 	CONFIG_FLASH_PART_TYPE,
 	CONFIG_FLASH_PART_SIZE,
-	CONFIG_FLASH_MINOR,
```
Comment (sunghan-chang) [inline os/board/common/partitions.c:43]: 1. Please remove CONFIG_FLASH_MINOR from all of defconfigs.
2. Please remove config FLASH_MINOR from os/board/common/Kconfig.

File: build/configs/artik053/cxxtest/defconfig [hunk 13]
```diff
@@ -233,21 +233,20 @@ CONFIG_ARTIK05X_BOOT_FAILURE_DETECTION=y
...
-CONFIG_ARTIK05X_AUTOMOUNT_USERFS_MOUNTPOINT="/mnt"
 # CONFIG_ARTIK05X_AUTOMOUNT_SSSRW is not set
 
 #
 # Board-Partition Options
 #
 CONFIG_ARCH_BOARD_HAVE_FLASH=y
 CONFIG_FLASH_PARTITION=y
-CONFIG_FLASH_MINOR=0
+"
```
Comment (sunghan-chang) [inline build/configs/artik053/cxxtest/defconfig:243]: There is typo. Please remove this.

File: os/board/common/partitions.c (hunk 12 above)
Comment (jeongchanKim) [reply os/board/common/partitions.c:43]: Done. Thanks.

**Review by sunghan-chang — APPROVED** [review]

**Pull request merged**